src/ddmd/link.d

```diff
@@ -981,62 +925,409 @@ public int runProgram()
+         */
+        const(char)* getUCRTLibPath(bool x64)
+        {
+            if (UCRTSdkDir && UCRTVersion)
+               return FileName.buildPath(UCRTSdkDir, "Lib", UCRTVersion, x64 ? r"ucrt\x64" : r"ucrt\x86");
+            return null;
+        }
+
+        /**
+         * get the path to the Windows SDK CRT libraries
+         * Params:
+         *   x64 = target architecture (x86 if false)
+         * Returns:
+         *   folder containing the Windows SDK libraries
+         */
+        const(char)* getSDKLibPath(bool x64)
+        {
+            if (WindowsSdkDir)
+            {
+                const(char)* arch = x64 ? "x64" : "x86";
+                auto sdk = FileName.combine(WindowsSdkDir, "lib");
+                if (FileName.exists(FileName.buildPath(sdk, WindowsSdkVersion, "um", arch, "kernel32.lib"))) // SDK 10.0
+                    return FileName.buildPath(sdk, WindowsSdkVersion, "um", arch);
+                else if (FileName.exists(FileName.buildPath(sdk, r"win8\um", arch, "kernel32.lib"))) // SDK 8.0
+                    return FileName.buildPath(sdk, r"win8\um", arch);
+                else if (FileName.exists(FileName.buildPath(sdk, r"winv6.3\um", arch, "kernel32.lib"))) // SDK 8.1
+                    return FileName.buildPath(sdk, r"winv6.3\um", arch);
+                else if (x64 && FileName.exists(FileName.buildPath(sdk, arch, "kernel32.lib"))) // SDK 7.1 or earlier
+                    return FileName.buildPath(sdk, arch);
+                else if (!x64 && FileName.exists(FileName.buildPath(sdk, "kernel32.lib"))) // SDK 7.1 or earlier
+                    return sdk;
+            }
+            return null;
+        }
+
+        // iterate through subdirectories named by SDK version in baseDir and return the
+        //  one with the largest version that also contains the test file
+        static const(char)* findLatestSDKDir(const(char)* baseDir, const(char)* testfile)
+        {
+            auto allfiles = FileName.combine(baseDir, "*");
+            static if (!is(WIN32_FIND_DATAA)) alias WIN32_FIND_DATAA = WIN32_FIND_DATA; // support dmd 2.068
+            WIN32_FIND_DATAA fileinfo;
+            HANDLE h = FindFirstFileA(allfiles, &fileinfo);
+            if (h == INVALID_HANDLE_VALUE)
+                return null;
+
+            char* res = null;
+            do
+            {
+                if (fileinfo.cFileName[0] >= '1' && fileinfo.cFileName[0] <= '9')
+                    if (res is null || strcmp(res, fileinfo.cFileName.ptr) < 0)
+                        if (FileName.exists(FileName.buildPath(baseDir, fileinfo.cFileName.ptr, testfile)))
+                        {
+                            const len = strlen(fileinfo.cFileName.ptr) + 1;
+                            res = cast(char*) memcpy(mem.xrealloc(res, len), fileinfo.cFileName.ptr, len);
+                        }
+            }
+            while(FindNextFileA(h, &fileinfo));
+            FindClose(h);
```

Comment:
`FindClose()` returns 0 on failure. Please check it. https://msdn.microsoft.com/en-us/library/windows/desktop/aa364413(v=vs.85).aspx

Reply:
I know, but it feels a bit harsh to fail while a sensible result is already obtained. I've changed it nevertheless.